overlord/snapstate/component_install_test.go

```diff
@@ -0,0 +1,229 @@
+	componentYaml := fmt.Sprintf(`component: %s+%s
+type: test
+version: 1.0
+`, snapName, compName)
+	compPath := snaptest.MakeTestComponent(c, componentYaml)
+	compf, err := snapfile.Open(compPath)
+	c.Assert(err, IsNil)
+
+	ci, err := snap.ReadComponentInfoFromContainer(compf)
+	c.Assert(err, IsNil)
+
+	return ci, compPath
+}
+
+func createTestSnapInfoForComponent(c *C, snapName string, snapRev snap.Revision, compName string) *snap.Info {
+	snapYaml := fmt.Sprintf(`name: %s
+type: app
+version: 1.1
+components:
+  %s:
+    type: test
+`, snapName, compName)
+	info, err := snap.InfoFromSnapYaml([]byte(snapYaml))
+	c.Assert(err, IsNil)
+	info.SideInfo = snap.SideInfo{RealName: snapName, Revision: snapRev}
+
+	return info
+}
+
+func createTestSnapSetup(info *snap.Info, flags snapstate.Flags) *snapstate.SnapSetup {
+	return &snapstate.SnapSetup{
+		Base:        info.Base,
+		SideInfo:    &info.SideInfo,
+		Channel:     info.Channel,
+		Flags:       flags.ForSnapSetup(),
+		Type:        info.Type(),
+		Version:     info.Version,
+		PlugsOnly:   len(info.Slots) == 0,
+		InstanceKey: info.InstanceKey,
+	}
+}
+
+func (s *snapmgrTestSuite) setStateWithOneComponent(c *C, snapName string,
+	snapRev snap.Revision, compName string, compRev snap.Revision) {
+	ssi := &snap.SideInfo{RealName: snapName, Revision: snapRev,
+		SnapID: "snapidididididididididididididid"}
+	csi := snap.NewComponentSideInfo(naming.NewComponentRef(snapName, compName), compRev)
+	snapstate.Set(s.state, snapName, &snapstate.SnapState{
+		Active: true,
+		Sequence: snapstatetest.NewSequenceFromRevisionSideInfos(
+			[]*snapstate.RevisionSideState{
+				snapstate.NewRevisionSideInfo(ssi,
+					[]*snap.ComponentSideInfo{csi})}),
+		Current: snapRev,
+	})
+}
+
+func (s *snapmgrTestSuite) TestInstallComponentPath(c *C) {
+	const snapName = "mysnap"
+	const compName = "mycomp"
+	_, compPath := createTestComponent(c, snapName, compName)
+	info := createTestSnapInfoForComponent(c, snapName, snap.R(1), compName)
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	csi := snap.NewComponentSideInfo(naming.ComponentRef{
+		SnapName: snapName, ComponentName: compName}, snap.R(33))
+	ts, err := snapstate.InstallComponentPath(s.state, csi, info, compPath,
+		snapstate.Flags{})
+	c.Assert(err, IsNil)
+
+	verifyComponentInstallTasks(c, compOptIsLocal, ts)
+	c.Assert(s.state.TaskCount(), Equals, len(ts.Tasks()))
+	// File is not deleted
+	c.Assert(osutil.FileExists(compPath), Equals, true)
+}
+
+func (s *snapmgrTestSuite) TestInstallComponentPathCompRevisionPresent(c *C) {
+	const snapName = "mysnap"
+	const compName = "mycomp"
+	snapRev := snap.R(1)
+	compRev := snap.R(7)
+	_, compPath := createTestComponent(c, snapName, compName)
+	info := createTestSnapInfoForComponent(c, snapName, snapRev, compName)
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	// Current component same revision to the one we install
+	s.setStateWithOneComponent(c, snapName, snapRev, compName, compRev)
```

Comment:
do we need to distinguish between the revision being present in the current snap revision vs other snap revision?

Reply:
I've added a test case for this